### PR TITLE
feat: display departure times and search time in CET, regardless of which timezone the users is located in

### PR DESCRIPTION
**Pull request by jonasbrunvoll**
Closes https://github.com/AtB-AS/kundevendt/issues/16880

### Background

We have received feedback from users outside the Central European Timezone (CET), that when searching for trips, the departure times are displayed in their local time and not in CET. The trips should always be displayed in CET to avoid confusion.  

### Proposed solution

Added functionality to display departure times and the time in the search time selector in CTE, regardless of which timezone the user is located in.

### Acceptance Criteria

- [ ] Regardless of timezone, the search selector should display CET time. 
- [ ] Regardless of timezone, the departure times should be displayed with CET time when departure time is set to now.
- [ ] Regardless of timezone, the departure times should be displayed with CET time when departure time is set to departure. Also attempt to change the search time.    
- [ ] Regardless of timezone, the departure times should be displayed with CET time when departure time is set to arrival. Also attempt to change the search time.  
  
### Test input

To test this feature, change the internal clock of the computer to the listed timezones. The departure times should and the search time selector should always be shown in CTE time. 
- [ ] Helsinki - Finland (+0200)
- [ ] Oslo - Norway (+0100)
- [ ] London - UK (+0000)
- [ ] Los Angeles - US (-0800)

### CODE REVIEW

**Comment by mikaelbr** [comment]
You didn't require any changes in how the date is listed in search results? That seems odd to me as dates are provided in UTC so printing local time on the client when providing from the server intuitively would be wrong, no?

File: src/utils/date.ts [hunk 1]
```diff
@@ -354,3 +357,17 @@ export function formatTripDuration(
 
   return { duration, departure, arrival };
 }
+
+export function setTimezoneIfNeeded(date: Date): Date {
+  if (Intl.DateTimeFormat().resolvedOptions().timeZone === CET) return date;
+  return new Date(date.toLocaleString(FALLBACK_LANGUAGE, { timeZone: CET }));
+}
+
+export function formatLocalTimeToCTE(localTime: number) {
```
[the remark this answers is not in the document]
Comment (jonasbrunvoll) [reply src/utils/date.ts:366]: Ops. Will fix👍

**Comment by adriansberg** [comment]
Don't we always operate with UTC dates, or are they converted to local date when parsed from ISO?

**Comment by mikaelbr** [comment]
When formatting as string the local time is used. As input in the date selector parsing without timezone is done so that will probably also use local timezone 🤔

**Comment by jonasbrunvoll** [comment]
setTimezoneIfNeeded is used in the search time selector to display in CET. When creating a tripQuery during a search in client/journey-planner, formatCETToLocal() formates the search time from CET to local time. Then, in trip() on the server side, the search time is again formated from local time to UTC before the search towards the db is performed.    

I am not exactly sure where the search results are formated from UTC to local time, but it could be when parsed from ISO. Therefore, to formate from local to CET before displaying the trip departure times, the setTimezoneIfNeeded() is implemented into formateToClock().

**Comment by adriansberg** [comment]
- `date-fns`'s `format` does not take any options for the formatting to be in a certain time zone I assume? 
- Not sure if I like how e.g. `formatCETToLocal` modifies the time by adding or subtracting to the original time based on the timezone offset. Is there no way to keep to UTC times and only alter output in views (date/time selector might be an exception)? The journey planner API takes in UTC times I would guess?
- Have this been testet by setting a different time zone on the machine? (Maybe that's my responsibility 🫣)

**Comment by jonasbrunvoll** [comment]
- No, `date-fns`'s  does not have an option to format to a certain timezone. 
- I can take a new look and investigate if it is possible to keep UTC times and only alter the views. As of now, the setTimezoneIfNeeded() in formatToClock() formates from local time to CET. I dont think how the timezone difference in the search time selector should be changed. 
- Yes, I have testet this with multiple different timezones on my machine locally and it seems to work. But, I do appreciate if others also would test this PR with different time zones😃

**Comment by mikaelbr** [comment]
In regards to testing with changed time zones, testing to vercel preview release is probably more realistic as locally it could be different with server time

**Comment by jonasbrunvoll** [comment]
# Timezone testing in preview
These are the test results when searching after trips from different timezones in preview. 
✅ means that the results from preview are the same as when searching frammr.no from a device using CTE. 
     
### CET (+0100) 
- Departure times in trip results when search = 'now' ✅
- Departure times in trip results when search = 'departure' and time = (current -1, current hour , current +1 hour) ✅
- Departure times detailed view ✅
- Departure times in trip result after navigating from detailed view ✅
- Departure times in search time selector after navigating from detailed view ✅  
### EET (+0200) 
- Departure times in trip results when search = 'now' ✅  
- Departure times in trip results when search = 'departure' and time = (current -1, current hour , current +1 hour) ✅
- Departure times detailed view ✅
- Departure times in trip result after navigating from detailed view ✅
- Departure times in search time selector after navigating from detailed view ✅   
### UTC (+0000) ✅
- Departure times in trip results when search = 'now' ✅ 
- Departure times in trip results when search = 'departure' and time = (current -1, current hour , current +1 hour) ✅
- Departure times detailed view ✅
- Departure times in trip result after navigating from detailed view ✅
- Departure times in search time selector after navigating from detailed view ✅  
### PST (-0800) ✅  
-  Departure times in trip results when search = 'now' ✅ 
- Departure times in trip results when search = 'departure' and time = (current -1, current hour, current +1 hour) ✅ 
- Departure times detailed view ✅ 
- Departure times in trip result after navigating from detailed view ✅ 
- Departure times in search time selector after navigating from detailed view ✅

**Review by adriansberg — CHANGES_REQUESTED** [review]
Let's add a couple test to make sure that we don't break anything.

**Review by adriansberg — APPROVED** [review]

**Comment by mikaelbr** [comment]
After the deployment action hook change for Playwright tests it takes longer for GH to discover playwright tests and run them. So they aren't run yet. Think it is good to wait for them to run

**Comment by adriansberg** [comment]
Oh, that's right. I just glanced at the checks to see if all were green 🫣

**Comment by jonasbrunvoll** [comment]
Is there some way to force GH to discover and run the playwright tests, or do we just wait until GH finds them?

**Comment by mikaelbr** [comment]
They are visible here: https://github.com/AtB-AS/planner-web/actions/workflows/playwright.yml

looks like it takes some time before it runs, so think we could look into if this can be improved somewhat

**Pull request merged**